File: CHTL/CHTLGenerator/Generator.cpp
#include "Generator.hpp"

namespace CHTL {

Generator::Generator() {}

std::string Generator::generate(const Program* program) {
    std::string html;
    for (const auto& node : program->children) {
        html += generateNode(node.get());
    }
    return html;
}

std::string Generator::generateNode(const Node* node) {
    if (const auto element = dynamic_cast<const ElementNode*>(node)) {
        return generateElementNode(element);
    }
    if (const auto text = dynamic_cast<const TextNode*>(node)) {
        return generateTextNode(text);
    }
    if (const auto comment = dynamic_cast<const CommentNode*>(node)) {
        return comment->toString();
    }
    return "<!-- Unknown Node Type -->";
}

std::string Generator::generateElementNode(const ElementNode* node) {
    // The 'text' tag is special; it doesn't create a <text> element.
    // It's just a container for a TextNode in the AST.
    if (node->tagName == "text") {
        if (!node->children.empty()) {
            // A well-formed 'text' element has one TextNode child.
            return generateNode(node->children[0].get());
        }
        return "";
    }

    std::string html;
    html += "<" + node->tagName;

    for (const auto& attr : node->attributes) {
        html += " " + attr.first + "=\"" + attr.second + "\"";
    }

    if (!node->styles.empty()) {
        html += " style=\"";
        for (const auto& style : node->styles) {
            html += style.first + ":" + style.second + ";";
        }
        html += "\"";
    }

    html += ">";

    for (const auto& child : node->children) {
        html += generateNode(child.get());
    }

    html += "</" + node->tagName + ">";
    return html;
}

std::string Generator::generateTextNode(const TextNode* node) {
    return node->value;
}

} // namespace CHTL

File: CHTL/CHTLGenerator/generator.py
import html
from CHTL.CHTLNode import nodes

class Generator:
    def __init__(self):
        self.self_closing_tags = {'area', 'base', 'br', 'col', 'embed', 'hr', 'img', 'input', 'link', 'meta', 'param', 'source', 'track', 'wbr'}

    def generate(self, ast: nodes.ProgramNode) -> str:
        return self._visit(ast)

    def _visit(self, node):
        method_name = f'_visit_{type(node).__name__.lower()}'
        visitor = getattr(self, method_name, lambda n: "")
        return visitor(node)

    def _visit_programnode(self, node: nodes.ProgramNode) -> str:
        return "".join(self._visit(child) for child in node.children)

    def _visit_elementnode(self, node: nodes.ElementNode) -> str:
        html_str = f"<{node.tag_name}"
        # Sort attributes for deterministic output
        for attr in sorted(node.attributes, key=lambda x: x.name):
            value = attr.value.strip("'\"")
            html_str += f' {attr.name}="{html.escape(value)}"'

        if node.tag_name in self.self_closing_tags and not node.children:
            return html_str + " />"

        html_str += ">"
        html_str += "".join(self._visit(child) for child in node.children)
        html_str += f"</{node.tag_name}>"
        return html_str

    def _visit_textnode(self, node: nodes.TextNode) -> str:
        return html.escape(node.value)
